### PR TITLE
Add documentation for `__builtin_object_size`.

### LINKED ISSUE #95635: __builtin_object_size suboptimal size when passed to function
Consider the following example:

```
#include <stdio.h>                                                                                                                                                                                                             
                                                                                                                                                                                                                     
static void                                                                                                                                                                                                                    
check_size(void *ptr)                                                                                                                                                                                                          
{                                                                                                                                                                                                                              
                                                                                                                                                                                                                               
    printf("%p, %zu\n", ptr, __builtin_object_size(ptr, 1));                                                                                                                                                                   
}                                                                                                                                                                                                                              
                                                                                                                                                                                                                               
int                                                                                                                                                                                                                            
main(int argc, char *argv[])                                                                                                                                                                                                   
{                                                                                                                                                                                                                              
    struct {                                                                                                                                                                                                                   
        unsigned char padding_l;                                                                                                                                                                                                     
        unsigned char buf[41];                                                                                                                                                                                                 
        unsigned char padding_r;                                                                                                                                                                                                     
    } __stack;                                                                                                                                                                                                                 
                                                                                                                                                                                                                               
    check_size(&__stack.buf);                                                                                                                                                                                                  
    printf("%p, %zu\n", &__stack.buf, __builtin_object_size(&__stack.buf, 1));                                                                                                                                                 
    return (0);                                                                                                                                                                                                                
}                                                                                                                                                                                                                              
```

When compiled with -O1 or higher, I observe the following:
```
0x806c11d6, 42
0x806c11d6, 41
```

That is, the `__builtin_object_size()` invocation in `check_size` returns a result as if it had been invoked as `__builtin_object_size(ptr, 0);` rather than `__builtin_object_size(ptr, 1);`.  This makes `_FORTIFY_SOURCE` a little more lenient than it should be in some weird corner cases.

### PR DESCRIPTION
**Pull request by zygoloid**
Explicitly describe how the Clang builtin works, given that it's not exactly the same as GCC's builtin of the same name -- but is drop-in compatible.

Fixes #95635.

### CODE REVIEW

File: clang/docs/LanguageExtensions.rst [hunk 1]
```diff
@@ -5368,16 +5368,85 @@ The ``#pragma comment(lib, ...)`` directive is supported on all ELF targets.
...
+
+For ``__builtin_object_size``, the value is determined entirely at compile time.
+With optimization enabled, better results will be produced, especially when the
+call to ``__builtin_object_size`` is in a different function from the formation
+of the pointer. Unlike in GCC, enabling optimization in Clang does not allow
+more information about subobjects to be determined, so the ``type & 1 == 1``
+case will often give imprecise results when used across a function call boundary
+even when optimization is enabled.
+
+`The ``pass_object_size`` and ``pass_dynamic_object_size`` attributes <https://clang.llvm.org/docs/AttributeReference.html#pass-object-size-pass-dynamic-object-size>`_
```
[the remark this answers is not in the document]
Comment (zygoloid) [reply clang/docs/LanguageExtensions.rst:5438]: Thanks.

```suggestion
`The pass_object_size and pass_dynamic_object_size attributes <https://clang.llvm.org/docs/AttributeReference.html#pass-object-size-pass-dynamic-object-size>`_
```

**Review by AaronBallman — APPROVED** [review]
Thank you for the additional docs, they LGTM!

**Comment by bwendling** [comment]
This LGTM as well. I have a question about the "sub-object" part. Do we attempt to discover the sub-object past any casting? A motivating example:

```c
char mux[10][10][10];

printf("%ld\n", __builtin_dynamic_object_size(&((char *)&mux[1][1])[var], 1));
```

GCC prints `89` (when `var` is 1). In the changes I'm making to the sub-object calculations, I want to say that this should print `9`, because the sub-object is `mux[1][1]` and not a one-dimensional array. There could be arguments made for both options, so I'd like to get everyone's opinions.

**Comment by zygoloid** [comment]
> I have a question about the "sub-object" part. Do we attempt to discover the sub-object past any casting?

Good question. I'm going to merge this PR, so I've filed #96737 for this ongoing discussion.

**Pull request merged**